### PR TITLE
Simplify for loops in fourslash.ts

### PR DESCRIPTION
**Pull request by andy-ms**

### CODE REVIEW

File: src/compiler/core.ts [hunk 1]
```diff
@@ -124,6 +124,13 @@ namespace ts {
         return undefined;
     }
 
+    export function zip<T, U>(arrayA: T[], arrayB: U[], callback: (a: T, b: U, index: number) => void): void {
```
Comment (sandersn) [inline src/compiler/core.ts]: lodash and ramda call this `zipWith`, and have `zip` with the type `(t: T[], u: U[]) => [T,U][]` (as do underscore, F#, Python and Haskell). I'd call it `zipWith`

File: src/harness/fourslash.ts [hunk 2]
```diff
@@ -692,15 +691,7 @@ namespace FourSlash {
...
-
-                let errorMsg = "\n" + "Member List contains: [" + members.entries[0].name;
-                for (let i = 1; i < members.entries.length; i++) {
-                    errorMsg += ", " + members.entries[i].name;
-                }
-                errorMsg += "]\n";
-
-                this.raiseError("Member list is not empty at Caret: " + errorMsg);
-
+                this.raiseError(`Member list is not empty at Caret:\nMember List contains: ${stringify(members.entries.map(e => e.name))}`);
```
Comment (sandersn) [inline src/harness/fourslash.ts]: are we using ES5's Array.map now?

File: src/harness/fourslash.ts [hunk 3]
```diff
@@ -710,13 +701,8 @@ namespace FourSlash {
...
             else if (completions && completions.entries.length !== 0 && !negative) {
-                let errorMsg = "\n" + "Completion List contains: [" + completions.entries[0].name;
-                for (let i = 1; i < completions.entries.length; i++) {
-                    errorMsg += ", " + completions.entries[i].name;
-                }
-                errorMsg += "]\n";
-
-                this.raiseError("Completion list is not empty at caret at position " + this.activeFile.fileName + " " + this.currentCaretPosition + errorMsg);
+                this.raiseError(`Completion list is not empty at caret at position ${this.activeFile.fileName} ${this.currentCaretPosition}\n` +
+                    `Completion List contains: ${stringify(completions.entries.map(e => e.name))}`);
```
Comment (sandersn) [inline src/harness/fourslash.ts]: same here. Are we using Array.map now?

Comment (sandersn) [inline src/harness/fourslash.ts]: I think it should be fine since fourslash only runs when testing the compiler, not when actually compiling.

Comment (andy-ms) [inline src/harness/fourslash.ts]: We can use `Array.map` now. We haven't been ES3-compatible for a while.

**Comment by mihailik** [comment]
Majority of changes replace `for` with `zip` with comparable LOC, slightly **worse** readability, potentially worse performance (due to closure allocations on heap).

It would make sense to run perf comparisons on various platforms — this change might be suboptimal for the reasons above.

**Comment by andy-ms** [comment]
This is test code -- it won't affect the time running the typescript compiler or services.

**Comment by mihailik** [comment]
Tests take a considerable time to run too.
And given that complexity didn't improve noticeably (arguably even gone up), what's the upside?

**Comment by sandersn** [comment]
1. Fourslash tests are slow because they do a lot of work -- probably too much. Micro-optimisations are unlikely to make a difference. Especially since only one or two of the changed functions is called more than a handful of times.
2. Readability is subjective, and the people who have to look at the fourslash code the most (@andy-ms and me, at least) find zip easier to read than for+indexing.
3. Tests take about 30 seconds to run on my machine. Compiling takes at least that long. I think the next build improvement should be compile time, not test time. That difference is largely because tests are parallel, so that's probably the best way forward for compiling too.

**Pull request merged**